text/0617-cloudfront-oac-l2.md

```diff
@@ -0,0 +1,877 @@
+      }
+
+      private grantDistributionAccessToBucket(distributionId: string): iam.AddToResourcePolicyResult {
+        const oacReadOnlyBucketPolicyStatement = new iam.PolicyStatement(
+          {
+            effect: iam.Effect.ALLOW,
+            principals: [new iam.ServicePrincipal('cloudfront.amazonaws.com')],
+            actions: ['s3:GetObject'],
+            resources: [this.bucket.arnForObjects('*')],
+            conditions: {
+              StringEquals: {
+                'AWS:SourceArn': `arn:${Aws.PARTITION}:cloudfront::${Aws.ACCOUNT_ID}:distribution/${distributionId}`,
+              },
+            },
+          },
+        );
+        const result = this.bucket.addToResourcePolicy(oacReadOnlyBucketPolicyStatement);
+        return result;
+      }
+
+      /**
+       * Use custom resource to update bucket policy
+       */
+      private grantDistributionAccessToImportedBucket(scope: Construct, distributionId: string) {
+        const provider = S3OriginAccessControlBucketPolicyProvider.getOrCreateProvider(scope, S3_ORIGIN_ACCESS_CONTROL_BUCKET_RESOURCE_TYPE,
+          {
+            description: 'Lambda function that updates S3 bucket policy to allow CloudFront distribution access.',
+          });
+        provider.addToRolePolicy({
+          Action: ['s3:getBucketPolicy', 's3:putBucketPolicy'],
+          Effect: 'Allow',
+          Resource: [this.bucket.bucketArn],
+        });
+
+        new CustomResource(scope, 'S3OriginBucketPolicyCustomResource', {
+          resourceType: S3_ORIGIN_ACCESS_CONTROL_BUCKET_RESOURCE_TYPE,
+          serviceToken: provider.serviceToken,
+          properties: {
+            DistributionId: distributionId,
+            AccountId: this.bucket.env.account,
+            Partition: Stack.of(scope).partition,
+            BucketName: this.bucket.bucketName,
+          },
+        });
+      }
+
+      /**
+       * Use custom resource to update KMS key policy
+       */
+      private grantDistributionAccessToKey(scope: Construct, distributionId: string, key: IKey) {
+        const provider = S3OriginAccessControlKeyPolicyProvider.getOrCreateProvider(scope, S3_ORIGIN_ACCESS_CONTROL_KEY_RESOURCE_TYPE,
+          {
+            description: 'Lambda function that updates SSE-KMS key policy to allow CloudFront distribution access.',
+          });
+        provider.addToRolePolicy({
+          Action: ['kms:PutKeyPolicy', 'kms:GetKeyPolicy', 'kms:DescribeKey'],
+          Effect: 'Allow',
+          Resource: [key.keyArn],
+        });
+
+        new CustomResource(scope, 'S3OriginKMSKeyPolicyCustomResource', {
+          resourceType: S3_ORIGIN_ACCESS_CONTROL_KEY_RESOURCE_TYPE,
+          serviceToken: provider.serviceToken,
+          properties: {
+            DistributionId: distributionId,
+            KmsKeyId: key.keyId,
+            AccountId: this.bucket.env.account,
+            Partition: Stack.of(scope).partition,
+          },
+        });
+      }
+    });
+  }
```

Comment:
TBH I glazed over everything between my last comment and this one because of the comment that I didn't think the custom resource was the way to go here.